content/chapters/compute/lecture/slides.mdpp

```diff
@@ -0,0 +1,1105 @@
+
+----
+
+### Copy-on-write
+
+When `fork` returns, the child duplicates the contents of the parent's page table.
+
+![Copy-on-write after fork()](./media/copy-on-write-initial.svg)
+
+----
+
+### Copy-on-write
+
+* Writable pages are **copied and modified** when written: `demo/copy-on-write/copy_on_write_overhead.c`
+* [Quiz](./quiz/sections-always-shared.md)
+
+![Copy-on-write after Write](./media/copy-on-write-final.svg)
+
+----
+
+### Copy-on-write Overheard
+
+* Writes cause minor page faults
+
+```
+student@os:~/.../compute/lecture/demo/copy-on-write$ ./copy_on_write_overhead 
+ * Child process started
+ -- Press ENTER to continue ...
+Time for reading 131072 pages: 30 ms
+ * Child process read pages
+ -- Press ENTER to continue ...
+Time for writing to 131072 pages: 222 ms
+[...]
+```
+
+* The child process performs exactly 131072 **minor** page faults when writing data
+
+```
+student@os:~$ cat /proc/`pidof copy_on_write_overhead`/stat | cut -d ' ' -f 10  # before writing
+22
+student@os:~$ cat /proc/`pidof copy_on_write_overhead`/stat | cut -d ' ' -f 10  # after writing
+131094
+```
+
+---
+
+## 2. Threads - Backstory
+
+* You have an application that applies filters to a set of images
+* Processes need to share all images
+    * Difficult because the images (like all data) are marked copy-on-write
+* **Solution:** use threads
+    * No copy-on-write: threads share the heap and data sections
+
+----
+
+### 2. Threads - Definitions
+
+* Each process has at least one thread
+* Lightweight processes (LWP)
+* Scheduling unit
+
+---
+
+### Thread - Resources
+
+![Thread Resources](media/process-vs-thread.png)
+
+----
+
+### App, Process, Thread - The Bigger Picture
+
+* [Quiz](./quiz/threads-shared-data.md)
+
+![Application Architecture](media/app-process-thread.svg)
+
+----
+
+### `fork` and Threads
+
+* `fork` also creates a new thread...
+* ... but part of a different process
+
+![fork - new Thread](./media/fork-thread.svg)
+
+---
+
+### Process vs Thread
+
+| PROCESS                          | THREAD                               |
+| :------------------------------- | :----------------------------------- |
+| independent                      | part of a process                    |
+| isolated VAS                     | shares VAS with other threads        |
+| slower creation                  | faster creation                      |
+| slower context switch            | shorter context switch               |
+| ending means ending all threads  | other threads continue when finished |
+
+----
+
+### Process vs Thread
+
+`demo/create-thread/create_thread.py`
+
+vs
+
+`demo/create-process/create_process.py`
+
+---
+
+## 3. Scheduling
+
+----
+
+### Scheduling - Why?
+
+* For multitasking: more threads than CPU cores
+* For responsiveness: small waiting time
+* For productivity: execute more threads
+
+----
+
+### Typical Thread Execution 
+
+![Thread Execution](media/thread-behaviour.png)
```

Comment:
We need to reference pictures from other sources.

Reply:
Any example how? It would probably be simpler to just make our own version of this. It would also be consistent with the other images we use.

Reply:
Add [footnotes](https://github.blog/changelog/2021-09-30-footnotes-now-supported-in-markdown-fields/).